elastic/elastic-agent
# Apply mTLS config from policy

## What does this PR do?
This PR implements reading and applying TLS configuration for Fleet client  using CA, certificate and key included in Fleet policy.

This PR:
- includes unit tests ported from #4398, 
- reimplements the solution of #4398 to adapt it to the new structure of internal/pkg/agent/application/actions/handlers/handler_action_policy_change.go
- modifies mock fleet server to support SSL configuration
- Add integration tests for both TLS and mTLS usecases using custom CAs

~~Note to reviewers: refactor of ProxyURL integration tests has been moved to PR #4813 , so for initial review you can have a look at [this set of commits](https://github.com/elastic/elastic-agent/pull/4770/files/cae577b7aa3ff22a4d9fbaeaff57ba5b71c088e1..bd9fe64bd86252f39f7b07f2354164fe85fbf860) or wait till PR #4813 is merged and this PR rebased onto the new main~~ 
PR #4813 has been merged and this change has been rebase on top of the new `main`.



## Why is it important?

Configuring TLS via the policy allows agent to connect to Fleet (possibly via a proxy) using custom CAs or enabling mTLS (certificate verification of both the client and the server).


## Checklist




## Disruptive User Impact



## How to test this PR locally

In order to test this PR we need:
- fleet server or proxy with TLS (or mTLS) configured using custom CAs and certificates signed by such CAs (left as exercise to the reader) 
- Make sure that the agent can connect to fleet for enrolling without custom CAs or certificates.
- enroll agent with a simple policy
- Add Custom CAs and or key + certificates for the agent to use along with a new URL (maybe a proxy) pointing to where TLS has been configured.
- Verify that the agent can connect correctly using CAs (and Certificate/Key in case of mTLS)



## Related issues


- Supersedes #4398
- Requires #4745
- Requires #4813 
- Closes #2247
- Closes #2248   

## Questions to ask yourself

- How are we going to support this in production?
- How are we going to measure its adoption?
- How are we going to debug this?
- What are the metrics I should take care of?
- ...
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
